spring-cloud-sleuth-zipkin/src/main/java/org/springframework/cloud/sleuth/zipkin/DiscoveryClientEndpointLocator.java

```diff
@@ -43,10 +46,22 @@ public Endpoint local() {
 		if (instance == null) {
 			throw new NoServiceInstanceAvailableException();
 		}
-		return Endpoint.create(instance.getServiceId(), getIpAddress(instance), instance.getPort());
+		return this.endpointCache.getEndpoint(createEndpointFactory(instance),
+				getHost(instance), instance.getPort(), instance.getServiceId());
 	}
 
-	private int getIpAddress(ServiceInstance instance) {
+	private EndpointCache.EndpointFactory createEndpointFactory(
+			final ServiceInstance instance) {
+		return new EndpointCacheImpl.EndpointFactory() {
```

Comment:
Declared in it's own `protected static` config class with a `@ConditionalOnProperty` then autowired (required=false).